### ISSUE TITLE
Traceback in Logs

### ISSUE THREAD

**Issue by datamel** [cylc/cylc-uiserver#308]
I am getting transient traceback on uiserver (all branches master), I think the steps below reproduce:

**Steps to reproduce the bug**

1. Have some pre-run (on cli) workflows in cylc-run directory (I had run rose test battery recently)

2. Start the hub: `cylc hub`.

I get traceback as follows:
```
[D 2022-02-11 10:18:29.780 CylcUIServer] sync_workflow(~<username>/rtb.20220211T094740Z/23-app-fcm-make-dest-cont-alt-with-share/-kpDDb)
[E 2022-02-11 10:18:29.788 CylcHubApp ioloop:761] Exception in callback functools.partial(<bound method IOLoop._discard_future_result of <tornado.platform.asyncio.AsyncIOMainLoop object at 0x7f76fb4c3e50>>, <Task finished name='Task-520' coro=<WorkflowsManager.update() done, defined at ~/cylc-uiserver/cylc/uiserver/workflows_mgr.py:224> exception=KeyError('req_client')>)
    Traceback (most recent call last):
      File “~/.conda/envs/cylc-8.0rc1.dev/lib/python3.9/site-packages/tornado/ioloop.py", line 741, in _run_callback
        ret = callback()
      File “~/.conda/envs/cylc-8.0rc1.dev/lib/python3.9/site-packages/tornado/ioloop.py", line 765, in _discard_future_result
        future.result()
      File “~/cylc-uiserver/cylc/uiserver/workflows_mgr.py", line 252, in update
        if not await self._connect(wid, flow):
      File “~/cylc-uiserver/cylc/uiserver/workflows_mgr.py", line 197, in _connect
        await self.uiserver.data_store_mgr.sync_workflow(
      File “~/cylc-uiserver/cylc/uiserver/data_store_mgr.py", line 154, in sync_workflow
        await self.entire_workflow_update(ids=[w_id])
      File "~/cylc-uiserver/cylc/uiserver/data_store_mgr.py", line 382, in entire_workflow_update
        gathers = [
      File "~/cylc-uiserver/cylc/uiserver/data_store_mgr.py", line 382, in <listcomp>
        gathers = [
      File "~ /cylc-uiserver/cylc/uiserver/data_store_mgr.py", line 377, in <genexpr>
        {'client': info['req_client'],
    KeyError: 'req_client'	
```

3. I then delete that folder and get a load of error messages like the following (One per workflow):

```
[D 2022-02-11 10:27:37.687 CylcUIServer] purge_workflow(~<username>/rtb.20220211T094740Z/06-app-prune-iso/N1etle)
[D 2022-02-11 10:27:37.717 CylcUIServer] unregister_workflow(~<username>/rtb.20220211T094740Z/06-app-prune-iso/N1etle)
[E 2022-02-11 10:27:37.718 CylcHubApp ioloop:761] Exception in callback functools.partial(<bound method IOLoop._discard_future_result of <tornado.platform.asyncio.AsyncIOMainLoop object at 0x7f76fb4c3e50>>, <Task finished name='Task-151537' coro=<WorkflowsManager.update() done, defined at ~/cylc-uiserver/cylc/uiserver/workflows_mgr.py:224> exception=KeyError('~<username>/rtb.20220211T094740Z/06-app-prune-iso/N1etle')>)
    Traceback (most recent call last):
      File “~/.conda/envs/cylc-8.0rc1.dev/lib/python3.9/site-packages/tornado/ioloop.py", line 741, in _run_callback
        ret = callback()
      File “~/.conda/envs/cylc-8.0rc1.dev/lib/python3.9/site-packages/tornado/ioloop.py", line 765, in _discard_future_result
        future.result()
      File “~/cylc-uiserver/cylc/uiserver/workflows_mgr.py", line 263, in update
        await self._unregister(wid)
      File “~/cylc-uiserver/cylc/uiserver/workflows_mgr.py", line 215, in _unregister
        await self.uiserver.data_store_mgr.unregister_workflow(wid)
      File “~/cylc-uiserver/cylc/uiserver/data_store_mgr.py", line 193, in unregister_workflow
        self.update_contact(w_id, pruned=True)
      File “~/cylc-uiserver/cylc/uiserver/data_store_mgr.py", line 117, in update_contact
        if 'delta_times' not in self.data[w_id]:
    KeyError: '~<username>/rtb.20220211T094740Z/06-app-prune-iso/N1etle'
[D 2022-02-11 10:27:38.018 CylcUIServer] purge_workflow(~<username>/rtb.20220211T094740Z/06-app-prune-iso/N1etle)
```
Happy to fix if someone can confirm the bug.

**Pull requests welcome!**
This is an Open Source project - please consider contributing a bug fix
yourself (please read `CONTRIBUTING.md` before starting any work though).

**Comment by MetRonnie** [comment]
Haven't seen the first traceback, but I have seen the second traceback recently - it should have been fixed by #306

**Comment by wxtim** [comment]
Sorry, I don't think that I can reproduce.

**Comment by datamel** [comment]
As discussed at team meeting, these errors have been reported by others although this has low priority as the errors have no impact on functionality of the ui.

**Comment by MetRonnie** [comment]
I think your working copy might not be up to date with master?
> ```
> File “~/cylc-uiserver/cylc/uiserver/data_store_mgr.py", line 193, in unregister_workflow
>        self.update_contact(w_id, pruned=True)
>```

That call is now on L194 on master

**Comment by oliver-sanders** [comment]
The `req_client` traceback is now explained in #311

**Comment by MetRonnie** [comment]
I have just had a similar traceback to the second for a deleted workflow
```
E 2022-02-15 16:12:58.391 CylcHubApp ioloop:761] Exception in callback functools.partial(
    <bound method IOLoop._discard_future_result of <tornado.platform.asyncio.AsyncIOMainLoop object at 0x7f31b8148580>>,
    <Task finished name='Task-353' coro=<WorkflowsManager.update() done, defined at
        ~/github/cylc-uiserver/cylc/uiserver/workflows_mgr.py:224> exception=KeyError(
            '~rdutta/cit-20220215T232302+0935/integration.test_examples/test_logging/e5230140-8e65-11ec-81fd-0050569a0f64')>)
    Traceback (most recent call last):
      File "~/.conda/envs/cylc8/lib/python3.9/site-packages/tornado/ioloop.py", line 741, in _run_callback
        ret = callback()
      File "~/.conda/envs/cylc8/lib/python3.9/site-packages/tornado/ioloop.py", line 765, in _discard_future_result
        future.result()
      File "~/github/cylc-uiserver/cylc/uiserver/workflows_mgr.py", line 267, in update
        self.active.pop(wid)
    KeyError: '~rdutta/cit-20220215T232302+0935/integration.test_examples/test_logging/e5230140-8e65-11ec-81fd-0050569a0f64'
```

**Comment by oliver-sanders** [comment]
That one's in #310

**Comment by oliver-sanders** [comment]
I think I have a solution to the `active.pop` and `KeyError: 'req_client'` tracebacks as a part of #311 & #312.

**Comment by oliver-sanders** [comment]
Of the three tracebacks reported here, two should have been fixed by #324, the third one:

```
KeyError: '~<username>/rtb.20220211T094740Z/06-app-prune-iso/N1etle'
```

May or may not have been fixed by #324, not sure.

**Comment by oliver-sanders** [comment]
Anyone seen the third traceback lately?

If not perhaps close this issue and re-open if tracebacks reoccur.

**Comment by datamel** [comment]
> Anyone seen the third traceback lately?
> If not perhaps close this issue and re-open if tracebacks reoccur.
I've not seen it in a while, happy to close and re-open if it re-surfaces.
